pkg/deployment/agency/cache.go

```diff
@@ -66,48 +69,135 @@ func (c cacheSingle) Data() (State, bool) {
 }
 
 type cache struct {
-	lock sync.Mutex
+	lock sync.RWMutex
 
 	valid bool
 
 	commitIndex uint64
 
 	data State
-}
 
-func (c *cache) CommitIndex() uint64 {
-	return c.commitIndex
+	leaderID string
 }
 
 func (c *cache) Data() (State, bool) {
-	c.lock.Lock()
-	defer c.lock.Unlock()
+	c.lock.RLock()
+	defer c.lock.RUnlock()
 
 	return c.data, c.valid
 }
 
-func (c *cache) Reload(ctx context.Context, client agency.Agency) (uint64, error) {
+// GetLeaderID returns a leader ID or empty string if a leader is not known.
+func (c *cache) GetLeaderID() string {
+	return c.leaderID
+}
+
+func (c *cache) Reload(ctx context.Context, clients []agency.Agency) (uint64, error) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
 
-	cfg, err := getAgencyConfig(ctx, client)
+	leaderCli, leaderConfig, err := getLeader(ctx, clients)
```

Comment:
If one of Agents is unreachable - then ctx will be already expired. Provide a lot smaller ctx for config discovery

Reply:
done